types.go

```diff
@@ -0,0 +1,302 @@
+package jsonschema
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strings"
+	"unicode"
+)
+
+type T int
+
+const (
+	Unknown T = iota
+	Null
+	Bool
+	Object
+	Array
+	Number
+	Integer
+	String
+)
+
+// MarshalJSON implements the json.Marshaler interface for T
+func (t T) MarshalJSON() ([]byte, error) {
+	return json.Marshal(mapTStr[t])
+}
+
+var mapTStr = [8]string{
+	Unknown: "unknown",
+	Null:    "null",
+	Bool:    "boolean",
+	Object:  "object",
+	Array:   "array",
+	Number:  "number",
+	Integer: "integer",
+	String:  "string",
+}
+
+var mapStrT = map[string]T{
+	"unknown": Unknown,
+	"null":    Null,
+	"boolean": Bool,
+	"object":  Object,
+	"array":   Array,
+	"number":  Number,
+	"integer": Integer,
+	"string":  String,
+}
+
+// dataToT resolves data to a JSON type.
+func dataToT(data interface{}) T {
+	if data == nil {
+		return Null
+	}
+
+	knd := reflect.TypeOf(data).Kind()
+	_ = knd
+
+	switch reflect.TypeOf(data).Kind() {
+	case reflect.Bool:
+		return Bool
+	case reflect.Float64:
+		number := reflect.ValueOf(data).Float()
+		if float64(int(number)) == number {
+			return Integer
+		}
+		return Number
+	case reflect.String:
+		return String
+	case reflect.Array, reflect.Slice:
+		return Array
+	case reflect.Map, reflect.Struct:
+		return Object
+	default:
+		return Unknown
+	}
+}
+
+// Ts is a collection of JSON types.
+type Ts []T
+
+func (ts Ts) String() string {
+	switch len(ts) {
+	case 0:
+		return ""
+	case 1:
+		return mapTStr[ts[0]]
+	}
+
+	// Calculate string length.
+	n := len(ts) - 1
+	for i := 0; i < len(ts); i++ {
+		n += len(mapTStr[ts[i]])
+	}
+
+	var b strings.Builder
```

Comment:
As the final length is knwon, the dynamic nature of `strings.Builder` doesn't seem necessary. Using directly a `[]byte` should be enough.